### PR TITLE
[fb-survey] Create codebook-generating tool

### PR DESCRIPTION
**Pull request by nmdefries**
### Description
Make tool to parse survey .qsf files and create a codebook with information about each survey item in a given wave.

### Changelog
- Create `generate-codebook.R` based off of [Jacob's Wave 7 processing code](https://github.com/jacobbien/covid-survey-processing).
- Create `qsf-utils.R` to share some basic utilities with `qsf-differ.R`.
- Create qsf tools directory and README.

### CODE REVIEW

File: facebook/qsf-tools/generate-codebook.R [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+  items <- displayed_questions %>% 
+    map_chr(~ .x$Payload$DataExportTag)
+  
+  # B13 was originally named incorrectly. Rename manually as needed
+  items[items == "B13 "] <- "B13"
+  
+  # get the text of the question:
+  questions <- displayed_questions %>% 
+    map_chr(~ .x$Payload$QuestionText) %>% 
+    str_remove_all("<[^<>]+>") %>%
```
Comment (capnrefsmmat) [inline facebook/qsf-tools/generate-codebook.R:53]: I realize that the survey only contains limited HTML and this is completely harmless here, but I can't resist linking this https://stackoverflow.com/a/1732454

**Review by capnrefsmmat — COMMENTED** [review]
Overall this looks good. I think the only thing I'd add is documenting the spreadsheets in the README, so operators adopting a new survey wave know which spreadsheets they need to update and how

File: facebook/qsf-tools/generate-codebook.R [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+                answers = answers,
+                randomization = randomization)
+
+  stopifnot(length(qdf$item) == length(unique(qdf$item)))
+  
+  # Remove blank questions (rare).
+  qdf <- qdf %>% filter(question != "Click to write the question text")
+  
+  # Add short question name mapped to item name
+  short_name_map <- c(
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply facebook/qsf-tools/generate-codebook.R]: I think you might be looking at an older version of the code. I've already changed the `map`s to be read from CSV.

File: facebook/qsf-tools/generate-codebook.R (hunk 1 above)
Comment (nmdefries) [reply facebook/qsf-tools/generate-codebook.R:53]: lol a classic :joy:

File: facebook/qsf-tools/generate-codebook.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (capnrefsmmat) [reply facebook/qsf-tools/generate-codebook.R]: Yup, I actually wrote that comment when you first opened the PR, then told you about it in a meeting, and GitHub kept the comment alive without any way for me to see it and remove it

**Comment by nmdefries** [comment]
>  the only thing I'd add is documenting the spreadsheets in the README

Do these exist yet?

**Comment by capnrefsmmat** [comment]
> Do these exist yet?

By "spreadsheet" I mean `item_replacement_map.csv`, `item_shortquestion_map.csv`, etc.

File: facebook/qsf-tools/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+.qsf, for example, which new survey questions replace which old survey
+questions. These mapping files are created manually and need to be updated for
+every new survey wave.
+
+* `item_replacement_map.csv`: Lists in-survey name of an `new_item` and the
+  in-survey name of the `old_item` it replaces. `new_item` should be the name
+  of a single item and be unique, but the `old_item` column has no formatting
+  requirements. It can hold a list of items, if the corresponding new survey
+  item is replacing multiple old questions, and a given item name can appear
+  in multiple rows of the `old_item` field.
```
Comment (capnrefsmmat) [inline facebook/qsf-tools/README.md:27]: "a list of items" means a comma-separated list, right?

**Review by capnrefsmmat — APPROVED** [review]
I think this looks good. I think the next steps are:

1. Get all the QSFs together
2. Run and produce a "prototype" codebook
3. Have Wichada and Facebook review it and suggest changes to the data (e.g. I'd probably add the timezone to the date column descriptions in `static_microdata_fields.csv`
4. Release to data users

**Pull request merged**